src/sort/map.rs

```diff
@@ -169,6 +174,45 @@ impl FromSort for ValueMap {
     }
 }
 
+struct MapRebuild {
```

Comment:
Can you use the existing interface for this? The `sort` interface has a `canonicalize` function that gets called for rebuilding. `canonicalize` is supposed to be more general than `find`, since it works for all of primitive values (e.g. i64), sorts, and containers.

The APIs surrounding canonicalization are a bit tangled (with `bad_find_value` which is now removed, `find`, `find_id`, and `canonicalize`). This is a good opportunity to clean up this.

Also, `canonicalize` implementations for `Set` and `Map` are removed in #176 but not for `Vec` I don't why...

Reply:
Yeah I would really love to, I couldn't figure out how to though. The problem is that the old `canonicalize` uses the union-find, while my new `rebuild` primitive needs to use the parent table. And I couldn't figure out how to get the parent table to `canonicalize`, so I made a separate code path.

any ideas on how to fix this?

Reply:
I think what this also implies is that `canonicalize` is broken in term mode. I suggest the following

1. Let `Sort::canonicalize` takes an e-graph and migrate terms mode canonicalization from `EGraph::find` to `Sort::canonicalize`.
2. Completely get rid of `EGraph::find`. Instead, always use `Sort::canonicalize`, so that we have a single source of truth.
3. Have a *single* `rebuild` primitive that accepts any value and calls `Sort::canonicalize` as its implementation.

Also, I think `rebuild` might not be a good name for this primitive, since rebuilding as defined in the EqSat literature means a different thing. I think `canonicalize` might be a better name.

Let me know if you have any questions or if this does not make sense to you. Happy to figure out a plan together.

Reply:
First a clarification: `canonicalize` isn't needed in term mode, so it doesn't matter that it's broken! I only use the `rebuild` primitives.
As for your plan, this was also my plan but I ran into a roadblock. As soon as you change `canonicalize` to take an egraph, you run into rust borrow errors that I can't resolve. I remember something about rebuilding needing a mutable egraph.

Reply:
Thanks for the clarification!